### PR TITLE
[4.0] Generate an autoload_psr4.php with extensions namespaces

### PR DESCRIPTION
**Pull request by yvesh**
### Summary of Changes

Generates an autoload_psr4.php file with namespace and folder for the namespaced extensions. (First version, there needs to be done some follow ups)

### Testing Instructions

Apply patch, and reload the page. Make sure there are no errors or notices and than check if there in the libraries folder there is a 'autoload_psr4.php' file created. (but only on first load)

This should be also updated on extension update, install, publish / unpublish etc.

As when the plugin is not enabled this can break the entire site, we should reevaluate protecting plugins / extensions OR move it to the installer etc. 

### Documentation Changes Required

Probably

### CODE REVIEW

File: libraries/namespacemap.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
+<?php
+/**
+ * @package    Joomla.Libraries
+ *
+ * @copyright  Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE
```
Comment (brianteeman) [inline libraries/namespacemap.php]: LICENSE.txt

File: libraries/namespacemap.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		if (!file_exists(self::$file))
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Check if the namespace mapping file is existing, if not create it
```
Comment (brianteeman) [inline libraries/namespacemap.php]: ==>
+	 * Check if the namespace mapping file exists, if not create it

File: libraries/namespacemap.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	/**
+	 * Path to the autoloader
+	 *
+	 * @var    string
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected static $file = JPATH_LIBRARIES . '/autoload_psr4.php';
+
+	/**
+	 * Check if the file is existing
```
Comment (brianteeman) [inline libraries/namespacemap.php]: ==>
Check if the file exists

File: libraries/namespacemap.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+			$content[] = "\t'" . $namespace . "'" . ' => array(JPATH_ROOT . ' . $pathString . '),';
+		}
+
+		$content[] = ');';
+
+		file_put_contents(self::$file, implode("\n", $content));
+	}
+
+	/**
+	 * Get the namespaced extensions out of the database
```
Comment (brianteeman) [inline libraries/namespacemap.php]: ==> from the database

**Comment by Bakual** [comment]
@wilsonge Honestly, this should never have been merged without any discussion. Especially since @yvesh himself said

> As when the plugin is not enabled this can break the entire site, we should reevaluate protecting plugins / extensions OR move it to the installer etc.

That's just not going to work using a plugin. Something as elementary as this has to be part of the extension installer itself.
Also it only takes care of components. But the methods retrieve all extensions....

**Comment by Bakual** [comment]
Also the autoload_psr4.php file needs to be added to .gitignore. 
It also likely needs the JEXEC or die statement and some comment about the purpose of the file would be nice as well.

**Comment by Bakual** [comment]
Plugin isn't even installed, the necessary SQL files are all missing.

**Comment by laoneo** [comment]
Agree, especially as the whole goal is not  clear from that pr where the journey should go.

**Comment by Bakual** [comment]
Created a PR to revert this one: https://github.com/joomla/joomla-cms/pull/16464

**Pull request merged**